File: src/main/java/fr/stardustenterprises/rust/wrapper/WrapperPlugin.java
// Copyright 2021 Arc'blroth
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package fr.stardustenterprises.rust.wrapper;

import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.tasks.TaskProvider;

/**
 * A plugin that wraps Rust's build systems,
 * for embedding Rust libraries in Java projects.
 */
@SuppressWarnings("unused")
public class WrapperPlugin implements Plugin<Project> {
    @Override
    public void apply(final Project project) {
        project.getConfigurations().create("default");

        WrapperExtension extension = project.getExtensions().create("rust", WrapperExtension.class);

        TaskProvider<WrapperTask> wrapperTask = project.getTasks().register("build", WrapperTask.class);
        project.afterEvaluate(__ -> {
            wrapperTask.get().configure(extension);
            project.getArtifacts().add("default", wrapperTask);
        });
    }
}

File: plugin/src/main/kotlin/fr/stardustenterprises/rust/wrapper/WrapperPlugin.kt
package fr.stardustenterprises.rust.wrapper

import fr.stardustenterprises.rust.common.RustPlugin
import fr.stardustenterprises.rust.wrapper.ext.WrapperExtension
import fr.stardustenterprises.rust.wrapper.task.CargoExecTask

class WrapperPlugin : RustPlugin() {
    override var pluginId: String = "fr.stardustenterprises.rust.wrapper"

    override fun applyPlugin() {
        val wrapperExtension = project.extensions.create("rust", WrapperExtension::class.java)

        this.project.afterEvaluate {
            it.run {
                tasks.create("", CargoExecTask::class.java).run {
                    this.group = "rust"
                    this.configure(wrapperExtension)
                }
            }
        }
    }

    override fun conflictsWithPlugins(): Array<String> =
        arrayOf("fr.stardustenterprises.rust.importer")
}